plugins/gimp/file-jxl-save.cc

```diff
@@ -10,122 +10,368 @@
+  pixels_buffer = g_malloc(buffer_size);
+
+  gegl_buffer_get(buffer, GEGL_RECTANGLE(0, 0, jxl_info.xsize, jxl_info.ysize),
+                  1.0, native_format, pixels_buffer, GEGL_AUTO_ROWSTRIDE,
+                  GEGL_ABYSS_NONE);
+
+  g_clear_object(&buffer);
+
+  // progress...
+  gimp_progress_update((gdouble)++cur_progress / (gdouble)max_progress);
+
+  // Multi-threaded parallel runner.
+  auto runner = JxlResizableParallelRunnerMake(nullptr);
+
+  JxlResizableParallelRunnerSetThreads(
+      runner.get(),
+      JxlResizableParallelRunnerSuggestThreads(jxl_info.xsize, jxl_info.ysize));
+
+  auto enc = JxlEncoderMake(/*memory_manager=*/nullptr);
+
+  if (JXL_ENC_SUCCESS != JxlEncoderSetParallelRunner(enc.get(),
+                                                     JxlResizableParallelRunner,
+                                                     runner.get())) {
+    g_printerr("JxlEncoderSetParallelRunner failed\n");
+    return false;
   }
 
-  CodecInOut io;
-
-  GeglBuffer* gegl_buffer = gimp_drawable_get_buffer(drawable_id);
-
-  // TODO(lode): is there a way to query whether the data type if float or int
-  // from gegl_buffer_get_format instead?
-  GimpPrecision precision = gimp_image_get_precision(image_id);
-  if (precision == GIMP_PRECISION_HALF_GAMMA) {
-    io.metadata.m.bit_depth.bits_per_sample = 16;
-    io.metadata.m.bit_depth.exponent_bits_per_sample = 5;
-  } else if (precision == GIMP_PRECISION_FLOAT_GAMMA) {
-    io.metadata.m.SetFloat32Samples();
-  } else {  // unsigned integer
-    // TODO(lode): handle GIMP_PRECISION_DOUBLE_GAMMA. 64-bit per channel is not
-    // supported by io.metadata.m.
-    const Babl* native_format = gegl_buffer_get_format(gegl_buffer);
-    uint32_t bits_per_sample = 8 *
-                               babl_format_get_bytes_per_pixel(native_format) /
-                               babl_format_get_n_components(native_format);
-    io.metadata.m.SetUintSamples(bits_per_sample);
+  if (JXL_ENC_SUCCESS != JxlEncoderSetBasicInfo(enc.get(), &jxl_info)) {
+    g_printerr("JxlEncoderSetBasicInfo failed\n");
+    return false;
   }
 
-  const GeglRectangle rect = *gegl_buffer_get_extent(gegl_buffer);
-  std::vector<float> pixel_data(rect.width * rect.height * (3 + has_alpha));
-  gegl_buffer_get(gegl_buffer, &rect, 1., format, pixel_data.data(),
-                  GEGL_AUTO_ROWSTRIDE, GEGL_ABYSS_NONE);
-  g_clear_object(&gegl_buffer);
-
-  JXL_RETURN_IF_ERROR(
-      read_buffer(rect.width, rect.height, pixel_data, std::move(icc), &io));
-  CompressParams params;
-  PassesEncoderState encoder_state;
-  PaddedBytes compressed;
-  ThreadPoolInternal pool;
-  params.butteraugli_distance = 1.f;
-  JXL_RETURN_IF_ERROR(EncodeFile(params, &io, &encoder_state, &compressed,
-                                 /*aux_out=*/nullptr, &pool));
-  JXL_RETURN_IF_ERROR(WriteFile(compressed, filename));
+  JxlColorEncoding color_encoding = {};
+  JxlColorEncodingSetToSRGB(&color_encoding,
```

Comment:
I feel that I must be missing something but shouldn’t we get the profile from GIMP?

Reply:
Yes, getting the profile from GIMP would make sense.  I either overlooked the old code that did that or didn't know how to do it.  I'll take a look at it and update the commit if I can figure it out.

Reply:
So my left hand doesn't know what my right hand is doing... there's some icc processing at the beginning of the file, but it doesn't go anywhere... Also looks like another CodecInOut access difference between Thread Pool Internal and Parallel Runner.  I'll poke around some more to see if there's something I overlooked earlier..

Reply:
ICC profiles should be working more or less as expected now.  If internal GIMP profiles are used in the image, then internal JPEG XL profiles will be embedded in the image.  Otherwise the user-specified ICC profile is embedded.

There are also some limits on when linear or perceptual gamma can be used.  Linear only works with float and perceptual only with int because images with linear int or perceptual float don't display properly.  Conversion by GIMP without any user intervention.